pkg/utils/lifecycle/lifecycle_test.go

```diff
@@ -0,0 +1,346 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package lifecycle
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/labels"
+
+	"sigs.k8s.io/kwok/pkg/apis/internalversion"
+	"sigs.k8s.io/kwok/pkg/utils/format"
+)
+
+func TestListAllPossibleStages(t *testing.T) {
+	stages := []*internalversion.Stage{
+		{
+			Spec: internalversion.StageSpec{
+				Selector: &internalversion.StageSelector{
+					MatchLabels: map[string]string{
+						"a": "b",
+						"c": "d",
+					},
+				},
+			},
+		},
+		{
+			Spec: internalversion.StageSpec{
+				Selector: &internalversion.StageSelector{
+					MatchAnnotations: map[string]string{
+						"e": "f",
+						"g": "h",
+					},
+				},
+			},
+		},
+	}
+	lc, err := NewLifecycle(stages)
+	if err != nil {
+		t.Fatal("Could not create a new lifecycle:", err)
+	}
+	label := labels.Set{
+		"a": "b",
+		"c": "d",
+	}
+	annotation := labels.Set{
+		"e": "f",
+		"g": "h",
+	}
+	var data interface{}
+	var possibleStages []*Stage
+	possibleStages, err = lc.ListAllPossible(label, annotation, data)
+	if err != nil {
+		t.Fatal("Could not list all possible Stages:", err)
+	}
+	if possibleStages == nil {
+		t.Fatal("Did not expect the list of all possible stages to be empty")
+	}
+	for i := 0; i < len(possibleStages); i++ {
+		stageLabels := possibleStages[i].matchLabels
+		stageAnnotation := possibleStages[i].matchAnnotations
+		if stageLabels != nil {
+			if stageLabels.String() != label.String() {
+				t.Fatal("StageLabels and label do not match")
+			}
+		}
+		if stageAnnotation != nil {
+			if stageAnnotation.String() != annotation.String() {
+				t.Fatal("StageAnnotations and annotation do not match")
+			}
+		}
+	}
+}
+
+func TestLifecycleMatch(t *testing.T) {
```

Comment:
Like the TestStageMatch, add various test data to cover the function as possible.